### PR TITLE
fix ambiguity error in is_intersection_empty for universal set

### LINKED ISSUE #2180: MethodError with disjointness check with universal set
```julia
julia> using Revise, LazySets

julia> x = CartesianProductArray([rand(Hyperrectangle) for _ in 1:3]);

julia> is_intersection_empty(Universe(2), x);
ERROR: MethodError: is_intersection_empty(::Universe{Float64}, ::CartesianProductArray{Float64,Hyperrectangle{Float64,Array{Float64,1},Array{Float64,1}}}) is ambiguous. Candidates:
  is_intersection_empty(U::Universe{N}, X::LazySet{N}) where N<:Real in LazySets at /home/mforets/.julia/dev/LazySets/src/ConcreteOperations/isdisjoint.jl:1184
  is_intersection_empty(P::AbstractPolyhedron{N}, cpa::CartesianProductArray{N,S} where S<:LazySet{N}) where N<:Real in LazySets at /home/mforets/.julia/dev/LazySets/src/ConcreteOperations/isdisjoint.jl:1329
Possible fix, define
  is_intersection_empty(::Universe{N}, ::CartesianProductArray{N,S} where S<:LazySet{N}) where N<:Real
Stacktrace:
 [1] top-level scope at REPL[3]:1
 [2] eval(::Module, ::Any) at ./boot.jl:331
 [3] eval_user_input(::Any, ::REPL.REPLBackend) at /home/mforets/Tools/julia-1.4.1-linux-x86_64/julia-1.4.1/share/julia/stdlib/v1.4/REPL/src/REPL.jl:86
 [4] run_backend(::REPL.REPLBackend) at /home/mforets/.julia/packages/Revise/tV8FE/src/Revise.jl:1165
 [5] top-level scope at REPL[1]:0
```

reported by @dfcaporale

### PR DESCRIPTION
**Pull request by dfcaporale**
Closes #2180.

### CODE REVIEW

File: src/ConcreteOperations/isdisjoint.jl [hunk 1]
```diff
@@ -1339,6 +1339,15 @@ function is_intersection_empty(hs::HalfSpace{N},
     return is_intersection_empty(cpa, hs)
 end
 
+function is_intersection_empty(cpa::CartesianProductArray{N,S} where S<:LazySet{N},
+                               ::Universe{N}) where {N<:Real}
+    return isempty(cpa)
```
Comment (mforets) [inline src/ConcreteOperations/isdisjoint.jl:1344]: hmm and if the `cpa` was empty, shouldn't `empty \cap empty` be true..? 

any thoughts on this implementation:

```julia
function is_intersection_empty(cpa::CartesianProductArray{N,S}, U::Universe{N}) where {N<:Real}
    @assert dim(cpa) == dim(U) "the dimensions should match but they are $(dim(cpa)) and $(dim(U)) respectively"

    return false
end

# symmetric case
function is_intersection_empty(U::Universe{N}, cpa::CartesianProductArray{N,S},) where {N<:Real}
    return is_intersection_empty(cpa, U)
end
```

Comment (schillic) [reply src/ConcreteOperations/isdisjoint.jl:1344]: > shouldn't empty \cap empty be true..?

Hm, good point, but this should apply to *any* set (not just CPA), right?

Comment (mforets) [reply src/ConcreteOperations/isdisjoint.jl:1344]: Indeed, the intersection of any lazy set `A` with the `EmptySet` should be empty. This doesn't seem to be implemented, eg.

```julia
julia> isdisjoint(rand(Hyperrectangle), EmptySet(2))
ERROR: MethodError: no method matching constraints_list(::EmptySet{Float64})
Closest candidates are:
  constraints_list(::HalfSpace{N,VN} where VN<:AbstractArray{N,1}) where N<:Real at /home/mforets/.julia/dev/LazySets/src/Sets/HalfSpace.jl:278
  constraints_list(::AbstractArray{N,2}, ::AbstractArray{N,1}) where N<:Real at /home/mforets/.julia/dev/LazySets/src/Sets/HalfSpace.jl:297
  constraints_list(::Interval{N,IN} where IN<:IntervalArithmetic.AbstractInterval{N}) where N<:Real at /home/mforets/.julia/dev/LazySets/src/Sets/Interval.jl:408
  ...
Stacktrace:
 [1] _normal_Vector(::EmptySet{Float64}) at /home/mforets/.julia/dev/LazySets/src/Sets/HalfSpace.jl:497
 [2] is_intersection_empty(::Hyperrectangle{Float64,Array{Float64,1},Array{Float64,1}}, ::EmptySet{Float64}, ::Bool; solver::GLPKMathProgInterface.GLPKInterfaceLP.GLPKSolverLP, algorithm::String) at /home/mforets/.julia/dev/LazySets/src/ConcreteOperations/isdisjoint.jl:968
 [3] is_intersection_empty at /home/mforets/.julia/dev/LazySets/src/ConcreteOperations/isdisjoint.jl:951 [inlined] (repeats 2 times)
 [4] top-level scope at REPL[7]:1
```

Comment (schillic) [reply src/ConcreteOperations/isdisjoint.jl:1344]: That is good. I was worried that we are unsound there :wink:

Comment (schillic) [reply src/ConcreteOperations/isdisjoint.jl:1344]: In fact the default implementation is correct.

https://github.com/JuliaReach/LazySets.jl/blob/cae43fb24de162ad660d7b946f7741aff45e457d/src/ConcreteOperations/isdisjoint.jl#L1182-L1190

Comment (mforets) [reply src/ConcreteOperations/isdisjoint.jl]: The last commit re-uses the code and checks that dimensions match; note that on master,

```julia
julia> using LazySets

julia> is_intersection_empty(Universe(2), rand(Hyperrectangle, dim=3))
false
```

**Comment by schillic** [comment]
Maybe add a test for the original issue?

File: src/ConcreteOperations/isdisjoint.jl (hunk 1 above)
Comment (schillic) [reply src/ConcreteOperations/isdisjoint.jl]: Good find, can you add a test for this?

File: test/unit_CartesianProduct.jl [hunk 2]
```diff
@@ -330,6 +330,11 @@ for N in [Float64, Float32]
           is_intersection_empty(cpa1_box, G_empty)
     @test !is_intersection_empty(cpa1, G) &&
           !is_intersection_empty(Approximations.overapproximate(cpa1), G)
+    @test !is_intersection_empty(cpa1, Universe(4)) &&
+          !is_intersection_empty(Universe(4), cpa2)
+    @test_throws AssertionError(is_intersection_empty(cpa1, Universe(3)))
```
Comment (mforets) [inline test/unit_CartesianProduct.jl:335]:
```suggestion
    @test_throws AssertionError is_intersection_empty(cpa1, Universe(3))
```

File: test/unit_CartesianProduct.jl [hunk 3]
```diff
@@ -330,6 +330,11 @@ for N in [Float64, Float32]
...
+    @test_throws AssertionError(is_intersection_empty(Universe(5), cpa2))
```
Comment (mforets) [inline test/unit_CartesianProduct.jl:336]:
```suggestion
    @test_throws AssertionError is_intersection_empty(Universe(5), cpa2)
```

**Pull request merged**